reconcile/terraform_cloudflare_users.py

```diff
@@ -0,0 +1,298 @@
+import sys
+from dataclasses import dataclass
+from typing import (
+    Any,
+    Iterable,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Set,
+    Tuple,
+)
+
+from reconcile import queries
+from reconcile.gql_definitions.terraform_cloudflare_users import (
+    terraform_cloudflare_roles,
+)
+from reconcile.gql_definitions.terraform_cloudflare_users.terraform_cloudflare_roles import (
+    CloudflareAccountRoleQueryData,
+    CloudflareAccountRoleV1,
+)
+from reconcile.status import ExitCodes
+from reconcile.utils import gql
+from reconcile.utils.external_resource_spec import ExternalResourceSpec
+from reconcile.utils.runtime.integration import QontractReconcileIntegration
+from reconcile.utils.secret_reader import SecretReader
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.terraform import safe_resource_id
+from reconcile.utils.terraform.config_client import (
+    ClientAlreadyRegisteredError,
+    TerraformConfigClientCollection,
+)
+from reconcile.utils.terraform_client import run_terraform
+from reconcile.utils.terrascript.cloudflare_client import (
+    AccountShardingStrategy,
+    IntegrationUndefined,
+    InvalidTerraformState,
+    TerrascriptCloudflareClientFactory,
+)
+from reconcile.utils.terrascript.models import (
+    CloudflareAccount,
+    Integration,
+    TerraformStateS3,
+)
+
+QONTRACT_INTEGRATION = "terraform_cloudflare_users"
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
+QONTRACT_TF_PREFIX = "qrtfcfusers"
+CLOUDFLARE_EMAIL_DOMAIN_ALLOW_LIST_KEY = "cloudflareEmailDomainAllowList"
+
+
+@dataclass
+class CloudflareUser:
+    email_address: str
+    account_name: str
+    org_username: str
+    roles: Set[str]
+
+
+class TerraformCloudflareUsers(QontractReconcileIntegration):
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION.replace("_", "-")
+
+    def get_early_exit_desired_state(
+        self, *args: Any, **kwargs: Any
+    ) -> Optional[dict[str, Any]]:
+
+        cloudflare_roles, settings = self._get_desired_state()
+
+        return {
+            "cloudflare_roles": cloudflare_roles.dict(),
+            CLOUDFLARE_EMAIL_DOMAIN_ALLOW_LIST_KEY: settings.get(
+                CLOUDFLARE_EMAIL_DOMAIN_ALLOW_LIST_KEY
+            ),
+        }
+
+    def _get_desired_state(
+        self,
+    ) -> Tuple[CloudflareAccountRoleQueryData, Mapping[str, Any]]:
+        cloudflare_roles = terraform_cloudflare_roles.query(
+            query_func=gql.get_api().query
+        )
+
+        settings = queries.get_app_interface_settings()
+        return cloudflare_roles, settings
+
+    def run(self, dry_run: bool, *args: Any, **kwargs: Any) -> None:
+        print_to_file = args[0]
+        account_name = args[1]
+        thread_pool_size = args[2]
+        enable_deletion = args[3]
+
+        cloudflare_roles, settings = self._get_desired_state()
+
+        secret_reader = SecretReader(settings=settings)
```

Comment:
We are trying to deprecate `SecretReader`. A type-safe alternative:

```python
from reconcile.utils.secret_reader import create_secret_reader
from reconcile.typed_queries.app_interface_vault_settings import (
    get_app_interface_vault_settings,
)


vault_settings = get_app_interface_vault_settings()
secret_reader = create_secret_reader(use_vault=vault_settings.vault)
```

You can also create a dedicated settings query for your integration that includes `cloudflareEmailDomainAllowList`, so you only need a single query instead of 2 different ones for the settings.